reframe/core/pipeline.py

```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
               more details.
 
         '''
-        if self.perf_patterns is None:
-            return
 
-        self._setup_perf_logging()
         with osext.change_dir(self._stagedir):
-            # Check if default reference perf values are provided and
-            # store all the variables tested in the performance check
-            has_default = False
-            variables = set()
-            for key, ref in self.reference.items():
-                keyparts = key.split(self.reference.scope_separator)
-                system = keyparts[0]
-                varname = keyparts[-1]
-                unit = ref[3]
-                variables.add((varname, unit))
-                if system == '*':
-                    has_default = True
-                    break
-
-            if not has_default:
-                if not variables:
-                    # If empty, it means that self.reference was empty, so try
-                    # to infer their name from perf_patterns
-                    variables = {(name, None)
-                                 for name in self.perf_patterns.keys()}
-
-                for var in variables:
-                    name, unit = var
-                    ref_tuple = (0, None, None, unit)
-                    self.reference.update({'*': {name: ref_tuple}})
-
-            # We first evaluate and log all performance values and then we
-            # check them against the reference. This way we always log them
-            # even if the don't meet the reference.
-            for tag, expr in self.perf_patterns.items():
-                value = sn.evaluate(expr)
-                key = '%s:%s' % (self._current_partition.fullname, tag)
-                if key not in self.reference:
-                    raise SanityError(
-                        "tag `%s' not resolved in references for `%s'" %
-                        (tag, self._current_partition.fullname))
+            if hasattr(self, '_rfm_perf_vars') or self._rfm_perf_fns:
+                if hasattr(self, 'perf_patterns'):
+                    raise ReframeSyntaxError(
+                        f"assigning a value to 'perf_pattenrs' conflicts ",
+                        f"with using the 'performance_variables' decorator ",
+                        f"(class {self.__class__.__qualname__})"
```

Comment:
> The only thing is that internally it would help if we have a single representation as we do now.

I don't remember what I exactly I meant with that, but I agree with the current design.